### PR TITLE
Enhance auth library

### PR DESCRIPTION
**Pull request by Flanker32**
- prompt user auth method, auth file location and subscription name
- check subscription while get azure client

### CODE REVIEW

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java [hunk 1]
```diff
@@ -7,51 +7,82 @@
...
 public class AzureClientFactory {
+
+    private static final String SUBSCRIPTION_TEMPLATE = "Subscription : %s(%s)";
+    public static final String SUBSCRIPTION_NOT_FOUND = "Subscription %s was not found in current account";
+    public static final String NO_AVAILABLE_SUBSCRIPTION = "No available subscription found in current account";
+
     public static Azure getAzureClient(AuthConfiguration auth, String subscriptionId) throws InvalidConfigurationException, IOException,
         AzureLoginFailureException, InterruptedException, ExecutionException {
-        AzureTokenCredentials azureTokenCredentials = AzureAuthHelper.getAzureTokenCredentials(auth);
+        AzureTokenCredentialsDecorator azureTokenCredentials = AzureAuthHelper.getAzureTokenCredentials(auth);
```
Comment (andxu) [inline azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: azureTokenWrapper

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java [hunk 2]
```diff
@@ -7,51 +7,82 @@
...
+            checkSubscription(azure, subscriptionId);
+            final Subscription subscription = azure.getCurrentSubscription();
+            Log.info(String.format(SUBSCRIPTION_TEMPLATE, subscription.displayName(), subscription.subscriptionId()));
+            return azure;
         }
 
         return null;
     }
+
+    private static void checkSubscription(Azure azure, String subscriptionId) throws AzureLoginFailureException {
```
Comment (andxu) [inline azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: shall we also support subscription name?

Comment (andxu) [reply azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: if subscriptionId is empty , should return the first one if there are only one

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java [hunk 3]
```diff
@@ -7,51 +7,82 @@
...
             if (StringUtils.isEmpty(subscriptionId) && AzureAuthHelperLegacy.isInCloudShell()) {
                 subscriptionId = AzureAuthHelperLegacy.getSubscriptionOfCloudShell();
             }
-            return StringUtils.isEmpty(subscriptionId) ?
+            final Azure azure = StringUtils.isEmpty(subscriptionId) ?
                     authenticated.withDefaultSubscription() :
-                        authenticated.withSubscription(subscriptionId);
-
+                    authenticated.withSubscription(subscriptionId);
+            checkSubscription(azure, subscriptionId);
```
Comment (andxu) [inline azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: if subscriptionId is empty and there is default subscription id from "azure cli" etc, should use this default id

File: azure-auth-helper/pom.xml [hunk 4]
```diff
@@ -8,7 +8,7 @@
         <relativePath>../azure-maven-plugins-pom</relativePath>
     </parent>
     <artifactId>azure-auth-helper</artifactId>
-    <version>0.2.0-SNAPSHOT</version>
+    <version>0.3.0-SNAPSHOT</version>
```
Comment (andxu) [inline azure-auth-helper/pom.xml]: we have not released 0.2.0, keep 0.2.0-SNAPSHOT?

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java (hunk 2 above)
Comment (Flanker32) [reply azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: Java SDK only support subscription id, and we could only auth with azure first and then get the subscription name.

Comment (Flanker32) [reply azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: Java SDK will use the first subscription id if no default subscription

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java (hunk 3 above)
Comment (Flanker32) [reply azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/auth/AzureClientFactory.java]: Will check default subscription and show warning if no subscription(default or pom) specified

File: azure-auth-helper/pom.xml (hunk 4 above)
Comment (Flanker32) [reply azure-auth-helper/pom.xml]: Fixed

**Comment by akaroml** [comment]
Can we do one thing in one PR?
"Enhance auth library" doesn't provide clear info of what this PR is doing exactly.

**Pull request closed without merging**